File: codec_header.rs
use byte::{BytesExt, BE};
use futures::{Stream, StreamExt};

fn type_string_serialize(out: &mut Vec<u8>, string: &str) {
    if string.len() > 255 { out.push(0); }
    else { out.push(string.len() as u8); for cha in string.chars() { out.push(cha as u8); } }
}
async fn type_string_deserialize<S: Stream<Item=u8>+Unpin>(stream: &mut S) -> Result<String,()> {
    let size = stream.next().await.ok_or(())?;
    let mut string = String::with_capacity(size as usize);
    for _ in 0..size { string.push(stream.next().await.ok_or(())? as char); }
    Ok(string)
}

fn type_float_serialize(out: &mut Vec<u8>, float: &f32) {
    let mut index = out.len();
    out.push(0); out.push(0); out.push(0); out.push(0);
    out.write_with::<f32>(&mut index, *float, BE);
}
async fn type_float_deserialize<S: Stream<Item=u8>+Unpin>(stream: &mut S) -> Result<f32, ()> {
    let buf = [
        stream.next().await.ok_or(())?,
        stream.next().await.ok_or(())?,
        stream.next().await.ok_or(())?,
        stream.next().await.ok_or(())?,
    ];
    buf.read_with(&mut 0, byte::BE).or(Err(()))
}

fn type_u16_serialize(out: &mut Vec<u8>, ushort: &u16) {
    let mut index = out.len();
    out.push(0); out.push(0);
    out.write_with::<u16>(&mut index, *ushort, byte::BE);
}
async fn type_u16_deserialize<S: Stream<Item=u8>+Unpin>(stream: &mut S) -> Result<u16, ()> {
    let buf = [
        stream.next().await.ok_or(())?,
        stream.next().await.ok_or(())?,
    ];
    buf.read_with(&mut 0, byte::BE).or(Err(()))
}

fn type_u32_serialize(out: &mut Vec<u8>, uint: &u32) {
    let mut index = out.len();
    out.push(0); out.push(0); out.push(0); out.push(0);
    out.write_with::<u32>(&mut index, *uint, byte::BE);
}
async fn type_u32_deserialize<S: Stream<Item=u8>+Unpin>(stream: &mut S) -> Result<u32, ()> {
    let buf = [
        stream.next().await.ok_or(())?,
        stream.next().await.ok_or(())?,
        stream.next().await.ok_or(())?,
        stream.next().await.ok_or(())?,
    ];
    buf.read_with(&mut 0, byte::BE).or(Err(()))
}

fn type_float_pair_serialize(out: &mut Vec<u8>, pair: &(f32, f32)) {
    type_float_serialize(out, &pair.0);
    type_float_serialize(out, &pair.1);
}
async fn type_float_pair_deserialize<S: Stream<Item=u8>+Unpin>(stream: &mut S) -> Result<(f32, f32), ()> {
    Ok((type_float_deserialize(stream).await?, type_float_deserialize(stream).await?))
}

fn type_u8_serialize(out: &mut Vec<u8>, ubyte: &u8) { out.push(*ubyte); }
async fn type_u8_deserialize<S: Stream<Item=u8>+Unpin>(stream: &mut S) -> Result<u8,()> {
    stream.next().await.ok_or(())
}

fn type_bool_serialize(out: &mut Vec<u8>, boolean: &bool) { out.push(if *boolean { 1 } else { 0 }); }
async fn type_bool_deserialize<S: Stream<Item=u8>+Unpin>(stream: &mut S) -> Result<bool,()> {
    stream.next().await.map(|val| val > 0).ok_or(())
}

File: codec_header.ts
export class Box<T> {
    v: T;
    constructor(v: T) { this.v = v; }
}

function type_string_serialize(out: number[], string: string) {
    if (string.length > 255) { out.push(0); }
    else {
        out.push(string.length);
        for (let i = 0; i < string.length; i++) out.push(string.charCodeAt(i));
    }
}
function type_string_deserialize(buf: Uint8Array, index: Box<number>): string {
    let out = "";
    const size = buf[index.v];
    index.v++;
    let i = index.v;	
    index.v += size;
    while (i < index.v) out += String.fromCharCode(buf[i++]);
    return out; 
}

function type_float_serialize(out: number[], float: number) {
    const arr = new Float32Array([float]);
    const view = new Uint8Array(arr.buffer);
    out.push(view[3], view[2], view[1], view[0]);
}
function type_float_deserialize(buf: Uint8Array, index: Box<number>): number {
    const arr = new Uint8Array([buf[index.v+3], buf[index.v+2], buf[index.v+1], buf[index.v]]);
    const view = new Float32Array(arr.buffer);
    index.v += 4;
    return view[0];
}

function type_ushort_serialize(out: number[], ushort: number) {
    const arr = new Uint16Array([ushort]);
    const view = new Uint8Array(arr.buffer);
    out.push(view[1], view[0]);
}
function type_ushort_deserialize(buf: Uint8Array, index: Box<number>): number {
    const arr = new Uint8Array([buf[index.v+1], buf[index.v]]);
    const view = new Uint16Array(arr.buffer);
    index.v += 2;
    return view[0];
}

function type_float_pair_serialize(out: number[], pair: [number, number]) {
    type_float_serialize(out, pair[0])
    type_float_serialize(out, pair[1]);
}
function type_float_pair_deserialize(buf: Uint8Array, index: Box<number>): [number, number] {
    return [type_float_deserialize(buf, index), type_float_deserialize(buf, index)];
}

function type_ubyte_serialize(out: number[], ubyte: number) { out.push(ubyte); }
function type_ubyte_deserialize(buf: Uint8Array, index: Box<number>): number { return buf[index.v++]; }

function type_boolean_serialize(out: number[], bool: boolean) { out.push(bool ? 1 : 0); }
function type_boolean_deserialize(buf: Uint8Array, index: Box<number>): boolean { return buf[index.v++] > 0; }